### PR TITLE
Bug fix: Jest unit tests

### PR DESCRIPTION
**Pull request by joeflack4**
## Changes
    Bug fix: Jest tests
    - Fixed a few syntax errors. Undefined reference errors mostly.
    - Fixed a value name that was changed: isoalted --> unlinked

    General
    - Update: setOp(): Refactored part of this function into new func decomposeToArray()

### CODE REVIEW

File: frontend/src/state/GraphState.jsx [hunk 1]
```diff
@@ -65,6 +65,7 @@ export class GraphContainer {
     //    it's also a list of paths; all the other specialConcepts are lists of concept_ids
     this.gd = graphData;  // concepts, specialConcepts, csmi, edges, concept_ids, filled_gaps,
                           // missing_from_graph, hidden_by_vocab, nonstandard_concepts_hidden
+    this.gd.specialConcepts = {};
     this.gd.specialConcepts.allButFirstOccurrence = [];
```
Comment (joeflack4) [inline frontend/src/state/GraphState.jsx:69]: ## Bug fix: `specialConcepts` undefined

File: frontend/src/state/GraphState.jsx [hunk 2]
```diff
@@ -573,8 +579,8 @@ function setOp(op, setA, setB) {
     difference: differenceWith,
     intersection: intersectionWith
   })[op];
-  if (setA instanceof Set || setA instanceof Iterator) setA = [...setA];
-  if (setB instanceof Set || setB instanceof Iterator) setB = [...setB];
+  setA = decomposeToArray(setA);
```
Comment (joeflack4) [inline frontend/src/state/GraphState.jsx:582]: ## Bug fix: `Iterator` undefined
Apparently, `Iterator` is [only native to Firefox](https://stackoverflow.com/a/12079454/5258518). [ChatGPT helped me refactor](https://chatgpt.com/share/d9555ff7-eece-44ee-8930-6eaa28493f0a).

I also refactored by abstracting to a function:

```js
function decomposeToArray(setOrIterator) {
  if (setOrIterator instanceof Set || (setOrIterator && typeof setOrIterator.next === 'function')) setOrIterator = [...setOrIterator];
  return setOrIterator;
}
```

File: frontend/src/state/GraphState.jsx
Comment (joeflack4) [inline frontend/src/state/GraphState.jsx:1]: ## Why errors only during tests?
I got these errors while running Jest tests, but when I use the app normally, they don't happen. I wonder why? @Sigfried Any ideas?

File: frontend/src/state/GraphState.jsx [hunk 3]
```diff
@@ -427,79 +428,86 @@ export class GraphContainer {
     let displayOrder = 0;
     let rows = {
       displayedRows: {
-        name: "Visible rows", displayOrder: displayOrder++,
+        name: "Visible rows",
```
Comment (joeflack4) [inline frontend/src/state/GraphState.jsx:431]: ## `name`, `displayOrder` on separate lines
I can change this back if you want; it's kinda arbitrary. ChatGPT liked this better and I know that having them on separate lines abides by code styles, and I'm undecided which I like better.

File: frontend/src/state/GraphState.jsx [hunk 4]
```diff
@@ -427,79 +428,86 @@ export class GraphContainer {
...
       },
       concepts: {
-        name: "Concepts", displayOrder: displayOrder++,
-        value: this.gd.concept_ids.length,
-        hiddenConceptCnt: setOp('difference', this.gd.concept_ids, displayedCids).length,
-        displayedConceptCnt: setOp('intersection', this.gd.concept_ids, displayedCids).length,
+        name: "Concepts",
+        displayOrder: displayOrder++,
+        value: this.gd.concept_ids?.length ?? 0,
+        hiddenConceptCnt: setOp('difference', this.gd.concept_ids ?? [], displayedCids).length,
```
Comment (joeflack4) [inline frontend/src/state/GraphState.jsx:439]: ## Edge case: Potentially undefined `concept_ids`
I [asked ChatGPT](https://chatgpt.com/share/64697079-8e8f-460d-8006-ee4040341645) to handle cases of potentially undefined `PROPERTY` in `this.gd.specialConcepts.PROPERTY.length?` patterns, but it was overzealous and handled other situations as well.
@Sigfried I leave it up to you if you want to keep this change or revert.

File: frontend/src/state/GraphState.jsx [hunk 5]
```diff
@@ -427,79 +428,86 @@ export class GraphContainer {
...
         specialTreatmentDefault: false,
       },
       definitionConcepts: {
-        name: "Definition concepts", displayOrder: displayOrder++,
-        value: this.gd.specialConcepts.definitionConcepts.length,
-        hiddenConceptCnt: setOp('difference', this.gd.specialConcepts.definitionConcepts, displayedCids).length,
+        name: "Definition concepts",
+        displayOrder: displayOrder++,
+        value: this.gd.specialConcepts.definitionConcepts?.length ?? 0,
+        hiddenConceptCnt: setOp('difference', this.gd.specialConcepts.definitionConcepts ?? [], displayedCids).length,
```
Comment (joeflack4) [inline frontend/src/state/GraphState.jsx:448]: ## Bug fix Potentially undefined special concepts types
I [asked ChatGPT](https://chatgpt.com/share/64697079-8e8f-460d-8006-ee4040341645) to handle cases of potentially undefined `PROPERTY` in `this.gd.specialConcepts.PROPERTY.length?` patterns.

@Sigfried I think that these are undefined because the code in `develop` is unfinished. I ran the test on the `add-concepts` branch, and though it did error out in some cases, it got farther than on `develop`, indicating to me that you have gotten further in your development of this part of handling special concepts.

@Sigfried I leave it up to you if you want to keep this change or revert. I think that when you've finished with the special concepts, this will no longer be necessary. It makes the code harder to understand, but it is also potentially useful edge case handling, though you can say much better than I how useful it is. I just know for now that without these changes, the Jest tests will fail.

File: frontend/src/index.test.js [hunk 6]
```diff
@@ -2505,27 +2505,25 @@ test('test makeGraph() - diagram case', () => {
...
+// });
 
 // - roots, leaves
 // todo?
 // test('test roots, leaves - diagram case', () => {});
 
 test('test roots, leaves - 1 cset "single small"', () => {
   const gc = new GraphContainer(singleSmallGraphContainerGraphData);
-  expect(gc.roots).toStrictEqual(["442793", "4044391", "4129519", "isolated"]);
+  expect(gc.roots).toStrictEqual(["442793", "4044391", "4129519", "unlinked"]);
```
Comment (joeflack4) [inline frontend/src/index.test.js:2521]: ## Bug fix: `isolated` --> `unlinked`
I needed to update this to what you had in `GraphState.jsx`. I wrote this on Feb 7, and you wrote this unlinked bit on Feb 4, so the code was probably rapidly changing, and our branches were out of sync.

```js
    this.graph.addNode('unlinked');
    this.nodes['unlinked'] = unlinkedConceptsParent;
    for (let c of this.unlinkedConcepts) {
      this.graph.addDirectedEdge('unlinked', c);
    }
```

@Sigfried However, I don't really understand what it's trying to say here. What does it mean to have a root called `unlinked`? I can understand a concept being unlinked, but it should still have a concept ID. Note that graph objects we're passing to `GraphContainer` don't have anything inside them called 'unlinked' or 'isolated'.

File: frontend/src/state/GraphState.jsx (hunk 5 above)
Comment (joeflack4) [reply frontend/src/state/GraphState.jsx:448]: Siggie showed me that `lodash` has a much easier to read solution, e.g.:
`get(this, "gd.concept_ids") || []`

File: frontend/src/index.test.js (hunk 6 above)
Comment (joeflack4) [reply frontend/src/index.test.js:2521]: `unlinked` becomes the root of all orphans. they get inserted below that
